packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart

```diff
@@ -46,36 +46,36 @@ class DiffPaneController extends DisposableController {
   // is taken, and is used to assign a unique id to each [SnapshotListItem].
   int _snapshotId = 0;
 
-  VoidCallback? takeSnapshotHandler(String gaEvent) {
-    if (_isTakingSnapshot.value) return null;
-    return () {
-      ga.select(
-        gac.memory,
-        gaEvent,
-      );
-      unawaited(takeSnapshot());
-    };
-  }
-
   Future<void> takeSnapshot() async {
     _isTakingSnapshot.value = true;
-    final snapshots = core._snapshots;
+    ga.select(
+      gac.memory,
+      gac.MemoryEvent.diffTakeSnapshotControlPane,
+    );
 
     final item = SnapshotInstanceItem(
       id: _snapshotId++,
       displayNumber: _nextDisplayNumber(),
       isolateName: selectedIsolateName ?? '<isolate-not-detected>',
     );
 
+    final snapshots = core._snapshots;
     snapshots.add(item);
 
-    final heapData = await snapshotTaker.take();
-    await item.initializeHeapData(heapData);
-
-    final newElementIndex = snapshots.value.length - 1;
-    core._selectedSnapshotIndex.value = newElementIndex;
-    _isTakingSnapshot.value = false;
-    derived._updateValues();
+    try {
+      final heapData = await snapshotTaker.take();
+      await item.initializeHeapData(heapData);
+
+      final newElementIndex = snapshots.value.length - 1;
+      core._selectedSnapshotIndex.value = newElementIndex;
+    } catch (e) {
+      snapshots.remove(item);
+      core._selectedSnapshotIndex.value = 0;
```

Comment:
instead of 0 should this be the last valid snapshot in the list? For example, if i was looking at snapshot 2, and then took snapshot 3, I'd expect the index after error to be snapshot 2

Reply:
in fact, if we move line 69 and 70 into the finally block, that would allow us to remove this line entirely.